wallet/src/Pos/Wallet/Web/Mode.hs

```diff
@@ -7,90 +7,127 @@ module Pos.Wallet.Web.Mode
+import           Pos.Update.Configuration         (HasUpdateConfiguration)
+import           Pos.Util                         (Some (..))
+import           Pos.Util.JsonLog                 (HasJsonLogConfig (..), jsonLogDefault)
+import           Pos.Util.LoggerName              (HasLoggerName' (..),
+                                                   getLoggerNameDefault,
+                                                   modifyLoggerNameDefault)
+import qualified Pos.Util.OutboundQueue           as OQ.Reader
+import           Pos.Util.TimeWarp                (CanJsonLog (..))
+import           Pos.Util.UserSecret              (HasUserSecret (..))
+import           Pos.Util.Util                    (postfixLFields)
+import           Pos.Wallet.Redirect              (MonadBlockchainInfo (..),
+                                                   MonadUpdates (..),
+                                                   applyLastUpdateWebWallet,
+                                                   blockchainSlotDurationWebWallet,
+                                                   connectedPeersWebWallet,
+                                                   localChainDifficultyWebWallet,
+                                                   networkChainDifficultyWebWallet,
+                                                   waitForUpdateWebWallet)
+import           Pos.Wallet.SscType               (WalletSscType)
+import           Pos.Wallet.Web.ClientTypes       (Addr, CHash, CId (..), cIdToAddress)
+import           Pos.Wallet.Web.Error             (WalletError (..))
+import           Pos.Wallet.Web.Sockets.ConnSet   (ConnectionsVar)
+import           Pos.Wallet.Web.State.State       (WalletState)
+import           Pos.Wallet.Web.Tracking          (MonadBListener (..), onApplyTracking,
+                                                   onRollbackTracking)
+import           Pos.WorkMode                     (RealModeContext (..))
 
 
 
 
 data WalletWebModeContext = WalletWebModeContext
     { wwmcWalletState     :: !WalletState
     , wwmcConnectionsVar  :: !ConnectionsVar
+    , wwmcHashes          :: !AddrCIdHashes
     , wwmcRealModeContext :: !(RealModeContext WalletSscType)
     }
 
+newtype AddrCIdHashes = AddrCIdHashes { unAddrCIdHashes :: (IORef (Map CHash Address)) }
+
 makeLensesWith postfixLFields ''WalletWebModeContext
 
+instance HasLens AddrCIdHashes WalletWebModeContext AddrCIdHashes where
+    lensOf = wwmcHashes_L
+
+cIdToAddress' :: (MonadWalletWebMode m) => CId w -> m Address
+cIdToAddress' = either (throwM . DecodeError) pure . cIdToAddress
+
+convertCIdTOAddr :: (MonadWalletWebMode m) => CId Addr -> m Address
+convertCIdTOAddr i@(CId id) = do
+    hmRef <- unAddrCIdHashes <$> view (lensOf @AddrCIdHashes)
+    hm <- liftIO $ readIORef hmRef
+    case id `M.lookup` hm of
+       Just addr -> return addr
+       _ -> cIdToAddress' i >>= \addr ->
+            liftIO (modifyIORef' hmRef (M.insert id addr)) $> addr
```

Comment:
you may call this function on multithreading environment, why didn't you use `atomicModifyIORef`?

Reply:
I asked the same, my assumption about @georgeee's motivation is that we don't care about lost addresses, but want to fill this map as fast as possible so use more lightweight `modifyIORef`

Reply:
Making it an atomic operation:
(1) you gain guarantee of cache to be updated always
(2) you introduce potential contention problem (when several threads would try to access the resource, they will hang on same lock around the cache)

For this cache we don't need (1) and better avoid (2)

Reply:
George, you realise we can do `atomicModifyIORef` so there's near-0 contention. And we can do bulk updates thanks to lazyness.